rspace-bench/src/test/scala/coop/rchain/rspace/bench/BasicBench.scala

```diff
@@ -1,72 +1,263 @@
+import coop.rchain.rspace._
 import coop.rchain.rspace.history.Branch
-import coop.rchain.rspace.util._
-import coop.rchain.rspace.{LMDBStore, _}
-import org.openjdk.jmh.annotations.{Benchmark, Scope, State, TearDown}
+import coop.rchain.catscontrib.TaskContrib.TaskOps
+import coop.rchain.crypto.hash.Blake2b512Random
+import coop.rchain.models.Expr.ExprInstance.{GInt, GString}
+import coop.rchain.models.TaggedContinuation.TaggedCont.ParBody
+import coop.rchain.rholang.interpreter.accounting.Cost
+import org.openjdk.jmh.annotations.{State => _, _}
 import coop.rchain.shared.PathOps.RichPath
+import monix.eval.Task
+import monix.execution.Scheduler.Implicits.global
+import org.openjdk.jmh.infra.Blackhole
+import org.scalacheck.{Arbitrary, Gen}
+import org.scalacheck.Gen.Parameters
+import org.scalacheck.rng.Seed
 
-import scala.concurrent.ExecutionContext.Implicits.global
+import scala.collection.immutable.{BitSet, Seq}
 
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@Fork(value = 1)
+@Warmup(iterations = 1)
+@Measurement(iterations = 10)
+@OperationsPerInvocation(value = 100)
 class BasicBench {
 
   import BasicBench._
 
   @Benchmark
-  def consumeProduce(state: BenchState): Unit = {
-
+  @BenchmarkMode(Array(Mode.AverageTime))
+  def consumeProduce(bh: Blackhole, state: BenchState): Unit = {
     val space = state.testSpace
+    for (i <- 0 to 100) {
+      val c1 = space
+        .consume(
+          state.channels(i) :: Nil,
+          state.data(i)._2 :: Nil,
+          state.tc.head,
+          false
+        )(state.matcher)
+        .unsafeRunSync
 
-    space
-      .consume(
-        List("ch1", "ch2"),
-        List(StringMatch("bad"), StringMatch("finger")),
-        new StringsCaptor,
-        false
-      )
+      assert(c1.right.get.isEmpty)
+      bh.consume(c1)
+
+      val r2 =
+        space.produce(state.channels(i), state.data(i)._1, false)(state.matcher).unsafeRunSync
 
-    val r1 = space.produce("ch1", "bad", false)
+      assert(r2.right.get.nonEmpty)
+      bh.consume(r2)
+    }
+    if (state.debug) {
+      assert(space.createCheckpoint().unsafeRunSync.log.size == 303)
+    }
+  }
 
-    assert(r1.right.get.isEmpty)
+  @Benchmark
+  @BenchmarkMode(Array(Mode.AverageTime))
+  def produceConsume(bh: Blackhole, state: BenchState): Unit = {
+    val space = state.testSpace
+    for (i <- 0 to 100) {
+      val r2 =
+        space.produce(state.channels(i), state.data(i)._1, false)(state.matcher).unsafeRunSync
 
-    val r2 = space.produce("ch2", "finger", false)
+      assert(r2.right.get.isEmpty)
+      bh.consume(r2)
 
-    runK(r2)
+      val c1 = space
+        .consume(
+          state.channels(i) :: Nil,
+          state.data(i)._2 :: Nil,
+          state.tc.head,
+          false
+        )(state.matcher)
+        .unsafeRunSync
 
-    assert(getK(r2).results.head.toSet == Set("bad", "finger"))
+      assert(c1.right.get.nonEmpty)
+      bh.consume(c1)
+    }
+    if (state.debug) {
```

Comment:
It would also make sense to add a (debug) check whether the tuplespace was emptied at the bottom of the for loop

Reply:
<o